src/Core/Services/Implementations/AzureQueueEventWriteService.cs

```diff
@@ -30,8 +33,44 @@ public async Task CreateAsync(IEvent e)
 
         public async Task CreateManyAsync(IList<IEvent> e)
         {
-            var json = JsonConvert.SerializeObject(e, _jsonSettings);
-            await _queueClient.SendMessageAsync(json);
+            if (!e?.Any() ?? true)
+            {
+                return;
+            }
+
+            if (e.Count == 1)
+            {
+                await CreateAsync(e.First());
+                return;
+            }
+
+            foreach(var json in SerializeMany(e))
+            {
+                await _queueClient.SendMessageAsync(json);
+            }
+        }
+
+        private IEnumerable<string> SerializeMany(IList<IEvent> events)
+        {
+            var strings = new List<string>();
```

Comment:
yeah, I agree.

I think you're on to something but the whole message needs to be valid json too, so we'd be building that list.

Maybe the reasonable way to go about it is to realize that the JSON serialization is probably pretty fast compared to wire times so serializing multiple time is likely OK.

What if I serialize each message, adding up to a total size, then serialize a `Take(n)` where `n` is the number that will fill the message? It's serializing each message object twice, but safer to let Newtonsoft handle the JSON for sure

Reply:
I was thinking you get a list of serialized messages, even with the length, and can use `.TakeWhile<T>(predicate)` with an accumulator on length that will "batch" them based on size, and serialize the overall `IEnumerable<T>` that results for each "batch" which will give you valid Array json.

Reply:
something like that anyway, but where you only serialize each message once, and then serialize each "batch" as a new array

Reply:
Yup, I think we're in agreement. I was saying you're serializing twice because each event object runs through the serializer twice (once to get the length, then another time to serialize the aggregated array.